cli/runtime/config/discovery_sources_node.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package config
+
+import (
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	configapi "github.com/vmware-tanzu/tanzu-framework/cli/runtime/apis/config/v1alpha1"
+	"github.com/vmware-tanzu/tanzu-framework/cli/runtime/config/nodeutils"
+)
+
+// DiscoveryType constants
+const (
+	DiscoveryTypeOCI        = "oci"
+	DiscoveryTypeLocal      = "local"
+	DiscoveryTypeGCP        = "gcp"
+	DiscoveryTypeKubernetes = "kubernetes"
+	DiscoveryTypeREST       = "rest"
+)
+
+// setDiscoverySources adds or updates the node discoverySources
+func setDiscoverySources(node *yaml.Node, discoverySources []configapi.PluginDiscovery) (persist bool, err error) {
+	opts := func(c *nodeutils.Config) {
+		c.ForceCreate = true
+		c.Keys = []nodeutils.Key{
+			{Name: KeyDiscoverySources, Type: yaml.SequenceNode},
+		}
+	}
+	discoverySourcesNode, err := nodeutils.FindNode(node, opts)
+	if err != nil {
+		return persist, err
+	}
+	for _, discoverySource := range discoverySources {
+		persist, err = setDiscoverySource(discoverySourcesNode, discoverySource)
+		if err != nil {
+			return persist, err
+		}
+	}
+	return persist, err
+}
+
+func setDiscoverySource(discoverySourcesNode *yaml.Node, discoverySource configapi.PluginDiscovery) (persist bool, err error) {
+	newNode, err := nodeutils.ConvertToNode[configapi.PluginDiscovery](&discoverySource)
+	if err != nil {
+		return persist, err
+	}
+	exists := false
+	var result []*yaml.Node
+	for _, discoverySourceNode := range discoverySourcesNode.Content {
+		discoverySourceType, discoverySourceName := getDiscoverySourceTypeAndName(discoverySource)
+		if discoverySourceType == "" || discoverySourceName == "" {
+			return false, errors.New("not found")
+		}
```

Comment:
Can this be moved out of the `for` loop?

Reply:
good catch. moved out of loop